### PR TITLE
feat(operators): add searchbar

### LINKED ISSUE #254: UX Improvement: Add instant search to Sidebar
To make is easy to find special operators, there should be an instant search on top of the operator list in the sidebar.

### PR DESCRIPTION
**Pull request by JWO719**
closes #254

### CODE REVIEW

**Comment by knittingcodemonkey** [comment]
It looks like there is a merge conflict here. Can you resolve that?

**Comment by JWO719** [comment]
thank you for your review. To be honest I would have expected several issues, but I'm totally fine with it :D

File: src/app/operators/operators.component.ts [hunk 1]
```diff
@@ -15,15 +15,33 @@ import {
...
 import { Subject } from 'rxjs/Subject';
-import { Observable } from 'rxjs/Observable';
-import { filter, takeUntil } from 'rxjs/operators';
+import {
+  filter,
+  takeUntil,
+  withLatestFrom,
+  flatMap,
+  map,
+  merge,
```
Comment (ashwin-sureshkumar) [inline src/app/operators/operators.component.ts]: There are imports here which are unused.

**Comment by ashwin-sureshkumar** [comment]
@JWO719 Also, while you filter down there is an empty space created for each of the category group that doesn't have any operators for the filtered value. 

screen shot 2018-02-20 at 9 09 33 am

**Comment by btroncone** [comment]
Just a style preference, but I would consider removing the floating placeholder for the search bar. I'm also not sure if we need the bottom line or if we could use the below header to frame it. What do you all think?

image

File: src/app/operators/operators.component.html [hunk 2]
```diff
@@ -5,12 +5,18 @@
...
+        <mat-label>Search</mat-label>
+        <input matInput value="" [formControl]="searchInput">
+        <mat-icon matSuffix>search</mat-icon>
+      </mat-form-field>
+    </form>
     <mat-nav-list *ngFor="let category of categories"
                   class="operator-list">
-      <h3 mat-subheader
-          class="category-subheader">{{ category }}</h3>
+      <h3 mat-subheader class="category-subheader" *ngIf="(groupedOperators[category] | filter: searchInput.value).length > 0">{{ category }}</h3>
```
Comment (jotatoledo) [inline src/app/operators/operators.component.html]: This wont have a good performance as the # of operators increase, as the filtering is run 2 times

File: src/app/operators/pipes/filter.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+@Pipe({
+  name: 'filter'
+})
+export class FilterPipe implements PipeTransform {
+  transform(items: any[], filter: string): any {
+    if (!items || !filter) {
+      return items;
+    }
+    return items.filter(item => item.name.indexOf(filter) !== -1);
```
Comment (jotatoledo) [inline src/app/operators/pipes/filter.pipe.ts]: I don´t like the pipe approach for this case.

Having no type is a two edged blade: doesn't tight the use of the pipe with a concrete type, but you lose type check at compile time.

File: src/app/operators/operators.component.html (hunk 2 above)
Comment (JWO719) [reply src/app/operators/operators.component.html]: yeah I had that idea too, but last time I was simply to supid to implement it that way. But you are definetly right, I will give it another try :D

**Comment by jsonberry** [comment]
@btroncone agreed on removal of floating placeholder and bottom line.

A UX concern I have is that the user is required to delete characters from search in order to reset the list. I think there should be an automatic clearing of the search if an operator is selected, and/or a reset button should be available.

Perhaps we could...
- Change placeholder to a traditional placeholder, use a color lighter than Material black, possibly a hint color
- Remove search icon
- Add reset button, where the search icon currently is, show only when the input is not empty

**Pull request closed without merging**